pkg/network/encoding/kafka.go

```
@@ -6,140 +6,111 @@
+	if e == nil {
 		return nil
 	}
 
-	encoder := &kafkaEncoder{
-		aggregations: make(map[types.ConnectionKey]*kafkaAggregationWrapper, len(payload.Conns)),
+	connectionData := e.byConnection.Find(c)
+	if connectionData == nil || len(connectionData.Data) == 0 || connectionData.IsPIDCollision(c) {
+		return nil
 	}
 
-	// pre-populate aggregation map with keys for all existent connections
-	// this allows us to skip encoding orphan Kafka objects that can't be matched to a connection
-	for _, conn := range payload.Conns {
-		for _, key := range network.ConnectionKeysFromConnectionStats(conn) {
-			if log.ShouldLog(seelog.TraceLvl) {
-				log.Tracef("Payload has a connection %v and was converted to kafka key %v", conn, key)
-			}
-			encoder.aggregations[key] = nil
-		}
-	}
-	encoder.buildAggregations(payload)
-	return encoder
+	return e.encodeData(connectionData)
 }
 
-func (e *kafkaEncoder) GetKafkaAggregations(c network.ConnectionStats) *model.DataStreamsAggregations {
+func (e *kafkaEncoder) Close() {
 	if e == nil {
-		return nil
+		return
 	}
 
-	connectionKeys := network.ConnectionKeysFromConnectionStats(c)
-	for _, key := range connectionKeys {
-		if aggregation := e.aggregations[key]; aggregation != nil {
-			return e.aggregations[key].ValueFor(c)
-		}
+	e.reset()
+	e.byConnection.Close()
+}
+
+func (e *kafkaEncoder) encodeData(connectionData *USMConnectionData[kafka.Key, *kafka.RequestStat]) []byte {
+	for _, kv := range connectionData.Data {
+		key := kv.Key
+		stats := kv.Value
+
+		kafkaAggregation := kafkaAggregationPool.Get().(*model.KafkaAggregation)
+
+		kafkaAggregation.Header.RequestType = uint32(key.RequestAPIKey)
+		kafkaAggregation.Header.RequestVersion = uint32(key.RequestVersion)
+		kafkaAggregation.Topic = key.TopicName
+		kafkaAggregation.Count = uint32(stats.Count)
+
+		e.aggregations.KafkaAggregations = append(e.aggregations.KafkaAggregations, kafkaAggregation)
 	}
-	return nil
+
+	serializedData, _ := proto.Marshal(e.aggregations)
```

Do we have a test that makes us comfortable with ignoring this error ?

---

1. in case of error, the output will be nil, so no impact on the payload
2. That's a legacy and happens in many locations in the code. We should explore changing it and checking the error, but it might cause performance issue, as this is an hot-code-path, so I'll defer that to another PR